src/NUnitFramework/framework/Internal/ExceptionHelper.cs

```diff
@@ -34,6 +36,11 @@ namespace NUnit.Framework.Internal
     /// </summary>
     public class ExceptionHelper
     {
+        /// <summary>
+        /// When true, each entry in Exception.Data is included in the exception message.
+        /// </summary>
+        public static bool OutputExceptionDataProperty { get; set; } = true;
```

Comment:
Sorry, I don't understand the first concern.  Are you saying that we want to include it by default and we shouldn't have this setting?

I was wondering about the parallel test runs.  I don't have a requirement for the feature to be optional so I'm happy to remove the setting if you think that's simpler all around.

Reply:
I'm saying a few things here...

1. We need to decide if we want this to be optional as a team. If we don't, then you shouldn't add any sort of feature switch.

2. If we do want it, then we have to decide whether the switch should be for the entire run, the assembly, the fixture or the individual test - or some combo of these.

3. If we do want it to be optional, then it can't be a static member and therefore can't be a member of this static class. It has to be handled elssewhere.

4. If we do want it, then I'd rather see this issue deal with implementing it up to the point where it can be used. That's what I meant by saying no part of NUnit was using the switch. I'm not saying that your PR has to go that far, unless you want it to, another PR could actually finish up and close the issue.

Reply:
What's the process for making these decisions?  Is this PR blocked until the decision is made or is there something I can do to make the PR mergeable until the decision is made?

Reply:
We wait for feedback from other team members. Shouldn't take long.

Meanwhile, how do you think it should be resolved? If it's an option how do you think the user should express it?

Reply:
In regards to parallel test runs: I don't see users of NUnit changing the configuration during a test run, so static configuration would be fine in all cases... except NUnit test runs.  Does NUnit have any pattern for this?  If not, we could add the property to the test context and disable it within the test run.

In regards to developers configuring the settings: I imagined it could be done in the AssemblySetup.   If we add the setting to TestContext, then the static setting could be the default with overrides possible per test run.

In the next iteration, I had planned to add the ability for developers to register delegates for appending additional data to the message.  This way we could add a delegate to include the properties on the exception.  We could also add details for framework or corporate specific exceptions.  The way I would use it at my company is to reuse a component that prints information to logs for easier debugging.  It will be helpful to get the same information in our tests.  The component also outputs the Data element which is why it would be useful to disable this behavior in the framework.

Reply:
@drewburlingame 
We have used statics in that way in the past but have ended up regretting it a few times, when users wanted to change it for particular test cases. Use of a static limits testability and it's really not that hard to avoid it. I feel that the static has to go.

If we use the TestContext, it should not be a static but an instance setting or perhaps simply a known property of tests. An instance setting that is never changed passes down to all the child tests, so you get the stable behavior without the risk of using a static.

In keeping with the principle of separating policy (whether to include Data) from implementation (the hellper class) I would lean toward making `includeData` a parameter of the method that builds the message and putting the decision at a higher level.

@nunit/framework-team These are the options I see

1. No opt-in/opt-out switch

2. Opt-in globally, using an argument passed by the runner as a PackageSetting.

3. Opt-in at each level of the test hierarchy, as we do for default floating point tolerance.

I think all of these are valid ideas but we should really tell @drewburlingame which way we want to go so he can finish this up!

Reply:
I'm pretty much in agreement, but I'm leading towards 1.
Are exceptions with a non-empty Data member that common? The ResultStateExceptions we throw won't have anything in .Data, so it would have to be an uncaught exception in NUnit user code. Even then, .Data is typically empty. Then, adding the opt-in/opt-out would be just in case a user in this rare situation dislikes the Data readout enough that it they'd want the ability to stop seeing it. I'm thinking we can just wait until we're asked for an opt-out feature.

Potentially there's a scenario I don't know about. Does this sound reasonable?

Reply:
@jnm2 I think that makes sense. Let's not do something until we need it. This is what YAGNI is really all about: We know we will most likely need something, but we don't know the exact shape of what we'll need, so we leave it till we have more info.